File: client/src/app/services/weight-log.service.ts
import {inject, Injectable} from "@angular/core";
import {HttpClient} from '@angular/common/http';
import {WeightLog} from '../models/User/user';
import {map, Observable} from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class WeightLogService {

  private http = inject(HttpClient);

  // Get weight logs for the last N days
  getWeightLogs(days: number = 30): Observable<WeightLog[]> {
    return this.http.get<any>(`/api/user/weight-logs?days=${days}`)
      .pipe(
        map(response => {
          const data = typeof response === 'string' ? JSON.parse(response) : response;
          return data.logs;
        })
      );
  }

  // Add a new weight log
  addWeightLog(weight: number, notes?: string): Observable<WeightLog> {
    return this.http.post<any>('/api/user/weight-logs', { weight, notes })
      .pipe(
        map(response => {
          const data = typeof response === 'string' ? JSON.parse(response) : response;
          return data.log;
        })
      );
  }

  // Update an existing weight log
  updateWeightLog(id: number, weight: number, notes?: string): Observable<WeightLog> {
    return this.http.put<any>(`/api/user/weight-logs/${id}`, { weight, notes })
      .pipe(
        map(response => {
          const data = typeof response === 'string' ? JSON.parse(response) : response;
          return data.log;
        })
      );
  }

  // Delete a weight log
  deleteWeightLog(id: number): Observable<any> {
    return this.http.delete<any>(`/api/user/weight-logs/${id}`)
      .pipe(
        map(response => {
          const data = typeof response === 'string' ? JSON.parse(response) : response;
          return data;
        })
      );
  }
}

File: server/src/main/java/vttp/final_project/services/userManagement/WeightLogService.java
package vttp.final_project.services.userManagement;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import vttp.final_project.models.userModels.WeightLog;
import vttp.final_project.repository.user.UserSqlRepository;
import vttp.final_project.repository.user.WeightLogRepository;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

@Service
public class WeightLogService {

    @Autowired
    private WeightLogRepository weightLogRepo;

    @Autowired
    private UserSqlRepository userSqlRepo;

    // Get weight logs for a user for the last X days
    public List<WeightLog> getWeightLogs(String email, int days) {
        Integer userId = userSqlRepo.getUserIdByEmail(email);
        if (userId == null) {
            throw new RuntimeException("User not found: " + email);
        }

        LocalDate startDate = LocalDate.now().minusDays(days);
        return weightLogRepo.getWeightLogs(userId, startDate);
    }

    // Add a new weight log or update existing for the current day
    public WeightLog addWeightLog(String email, BigDecimal weight, String notes) {
        Integer userId = userSqlRepo.getUserIdByEmail(email);
        if (userId == null) {
            throw new RuntimeException("User not found: " + email);
        }

        LocalDate today = LocalDate.now();
        // Check if there's already a log for today
        WeightLog existingLog = weightLogRepo.getWeightLogForDate(userId, today);

        if (existingLog != null) {
            // Update existing log
            existingLog.setWeight(weight);
            existingLog.setNotes(notes);
            weightLogRepo.updateWeightLog(existingLog);
            return existingLog;
        } else {
            // Create a new log
            WeightLog newLog = new WeightLog();
            newLog.setUserId(userId);
            newLog.setWeight(weight);
            newLog.setDate(today);
            newLog.setNotes(notes);
            return weightLogRepo.saveWeightLog(newLog);
        }
    }

    // Update an existing weight log
    public WeightLog updateWeightLog(String email, Long logId, BigDecimal weight, String notes) {
        Integer userId = userSqlRepo.getUserIdByEmail(email);
        if (userId == null) {
            throw new RuntimeException("User not found: " + email);
        }
        WeightLog log = weightLogRepo.getWeightLogById(logId, userId);
        if (log == null) {
            throw new RuntimeException("Weight log not found");
        }
        log.setWeight(weight);
        log.setNotes(notes);
        boolean updated = weightLogRepo.updateWeightLog(log);
        if (!updated) {
            throw new RuntimeException("Failed to update weight log");
        }

        return log;
    }

    // Delete a weight log
    public boolean deleteWeightLog(String email, Long logId) {
        Integer userId = userSqlRepo.getUserIdByEmail(email);
        if (userId == null) {
            throw new RuntimeException("User not found: " + email);
        }
        return weightLogRepo.deleteWeightLog(logId, userId);
    }


}
